src/hypofuzz/dashboard.py

```diff
@@ -19,32 +24,160 @@
+
+    async def initial(
+        self, reports: dict[str, list[Report]], metadata: dict[str, Metadata]
+    ) -> None:
+        # match same shape as overview for now (mapping of node id to data)
+        # TODO refactor this, drop the nesting and fix the types typescript-side
+        await self.send_json(
+            {
+                "type": "initial",
+                "reports": {self.node_id: reports[self.node_id]},
+                "metadata": {self.node_id: metadata[self.node_id]},
+            }
+        )
+
+    async def on_event(self, event_type: str, data: Any) -> None:
+        if event_type == "save":
+            node_id = data["nodeid"]
+            if node_id != self.node_id:
+                return
+            await self.send_json({"type": "save", "data": data})
+
+        if event_type == "metadata":
+            node_id = data["nodeid"]
+            if node_id != self.node_id:
+                return
+            await self.send_json({"type": "metadata", "data": data})
+
+        # TODO handle delete and refresh events
+
+
+REPORTS: dict[str, SortedList[Report]] = defaultdict(
+    lambda: SortedList(key=lambda r: r["elapsed_time"])
+)
+METADATA: dict[str, Metadata] = {}
 PYTEST_ARGS: Optional[list[str]] = None
-active_connections: set[WebSocket] = set()
+websockets: set[HypofuzzWebsocket] = set()
+delete_debounce = 300
+refreshed_at: dict[bytes, float] = defaultdict(int)
 
 
 async def websocket(websocket: WebSocket) -> None:
+    websocket = (
+        TestWebsocket(websocket, node_id)
+        if (node_id := websocket.query_params.get("node_id"))
+        else OverviewWebsocket(websocket)
+    )
     await websocket.accept()
-    active_connections.add(websocket)
+    websockets.add(websocket)
+    await websocket.initial(REPORTS, METADATA)
+
     try:
         while True:
             await websocket.receive_json()
     except WebSocketDisconnect:
-        active_connections.remove(websocket)
-
-
-async def broadcast_update(data: dict) -> None:
-    # copy since we might modify it
-    for connection in set(active_connections):
-        try:
-            await connection.send_json(data)
-        except WebSocketDisconnect:
-            active_connections.remove(connection)
+        websockets.remove(websocket)
+
+
+async def broadcast_event(event_type: str, data: dict) -> None:
+    for websocket in websockets:
+        await websocket.on_event(event_type, data)
+
+
+def database_listener(event: ListenerEventT) -> None:
+    event_type, args = event
+    if event_type == "save":
+        key, value = args
+        assert value is not None
+        if key.endswith(reports_key):
+            report = json.loads(value)
+            REPORTS[report["nodeid"]].add(report)
+            trio.run(broadcast_event, "save", report)
+        if key.endswith(metadata_key):
+            metadata = json.loads(value)
+            METADATA[metadata["nodeid"]] = metadata
+            trio.run(broadcast_event, "metadata", metadata)
```

Comment:
I'm pretty suspicious of these `trio.run` calls - I think it actually is safe in this case, because we're just sending some websocket events without other interaction, but it's inelegant to turn an async operation into a blocking sync operation.

Instead: we can `trio.open_memory_channel(inf)`, and `db.add_listener(send_chan.send_nowait)`; then start_soon a task in the `run_dashboard` nursery which does `async for event in recv_chan: <this handling logic>`.

Reply:
great point - I should have mentioned that I am not that good at async and so critiques of trio usages are very welcome. (I didn't even know open_memory_channel was a thing). I did see some trio failures if you refresh the dashboard fast enough which may well be from this call...

Reply:
Yep, that seems pretty likely to me!  

IME Trio is good at making easy things easy, and we should be pretty solidly in the easy-things-only regime for our little webapp.  As a general tip though, idiomatic use often leans heavily on context managers to ensure cleanup, and channels to communicate between tasks or act as a buffer (where that's necessary - and mostly here it's not).  Avoid async generators; see PEP-789 for details if curious.

Reply:
I'm still working through this one, but in case you have ideas - `watchdog` uses a background thread to track events, and `get_db()._db.add_listener(send_channel.send_nowait)` complains with `RuntimeError: must be called from async context` when that thread tries to send an event

Reply:
Try `partial(trio.from_thread.run_sync, send_chan.send_nowait, trio_token=trio.lowlevel.current_trio_token())` ([docs](https://trio.readthedocs.io/en/stable/reference-core.html#trio.from_thread.run_sync))